File: ruby_version/app/controllers/filter_controller.rb
class FilterController < ApplicationController
  
  def hash_filter
    
    @hash_filter = params[:hash_filter]
    
    @s = Search.where("md5hash = ?", @hash_filter).first
    
    @strs = "select image_id, count(*) from exif_data inner join images on exif_data.image_id = images.id where ";
    
    logger.info @s.serial.length
    
    i = 0
    @s.serial.each do |t|
      if i != 0
        @strs += " OR "
      end
      @strs += "(tag = '#{t["tag"]}' and value = '#{t["value"]}') "
      i = 1
    end
    
      @strs += "group by image_id having count(*) = #{@s.serial.length} "
    
      logger.info @strs
    ##array for the id's for searching later
    @id_array = Array.new
    
    @exif = ExifDatum.find_by_sql(@strs.to_s)
    @exif.each do |img|
      #logger.info img.image_id
      #logger.info img.image.location
      @id_array.push img.image_id
    end
    
    @exi2 = ExifDatum.select("tag, value, count(*) as count").where("image_id IN (?)", @id_array).group("tag, value").order("tag asc")
    #@exi2 = ExifDatum.all(:conditions => {["image_id = ?", @id_array]}) #.select("tag, value, count(*) as count") ##.group("tag, value").order("tag asc")
      @exi2.each do |exi2_data|
      
        @h = Array.new
        @s.serial.each do |t|
          @h.push({ "tag" => t["tag"], "value" => t["value"] })
        end
        @h.push({ "tag" => exi2_data.tag, "value" => exi2_data.value })

        logger.info @h
        logger.info "\n\n###"
        
        @y = @h.to_yaml
        @q = Digest::MD5.new.update(@y)

        filter_check(@q.to_s, @h)
        exi2_data.q = @q
        
      end
    
  end
  
end

File: Sources/App/Controllers/FilterController.swift
import Vapor
import Fluent
import SQLKit

/// GET /filter/:hash_filter — show images matching a saved search (set of
/// tag/value pairs), plus the tag/value counts within that result set so the
/// user can drill further. Mirrors FilterController#hash_filter.
///
/// SECURITY: the original Rails code interpolated tag/value strings directly
/// into raw SQL (`(tag = '#{t["tag"]}' and value = '#{t["value"]}')`), which is
/// SQL-injectable. This rewrite uses parameterized SQLKit bindings instead.
struct FilterController {
    func hashFilter(req: Request) async throws -> View {
        guard let hash = req.parameters.get("hash_filter") else {
            throw Abort(.badRequest, reason: "hash_filter is required")
        }

        guard let search = try await Search.query(on: req.db)
            .filter(\.$md5hash == hash)
            .first() else {
            throw Abort(.notFound, reason: "No saved search for hash \(hash)")
        }

        let pairs = search.tagValues
        let sql = req.db as! SQLDatabase

        // Find image_ids that have ALL of the requested tag/value pairs.
        // SELECT image_id FROM exif_data
        //   INNER JOIN images ON exif_data.image_id = images.id
        //   WHERE (tag = $1 AND value = $2) OR (...)
        //   GROUP BY image_id HAVING count(*) = N
        var imageIDs: [Int] = []
        if !pairs.isEmpty {
            // Build a parameterized OR group of (tag = ? AND value = ?) clauses.
            var orExprs: [SQLExpression] = []
            for pair in pairs {
                let clause = SQLBinaryExpression(
                    left: SQLBinaryExpression(
                        left: SQLColumn("tag"),
                        op: SQLBinaryOperator.equal,
                        right: SQLBind(pair.tag)
                    ),
                    op: SQLBinaryOperator.and,
                    right: SQLBinaryExpression(
                        left: SQLColumn("value"),
                        op: SQLBinaryOperator.equal,
                        right: SQLBind(pair.value)
                    )
                )
                orExprs.append(SQLGroupExpression(clause))
            }
            let whereExpr = orExprs.dropFirst().reduce(orExprs[0]) { acc, next in
                SQLBinaryExpression(left: acc, op: SQLBinaryOperator.or, right: next)
            }

            let rows = try await sql.select()
                .column("image_id")
                .from("exif_data")
                .join("images", on: "exif_data.image_id", .equal, "images.id")
                .where(whereExpr)
                .groupBy("image_id")
                .having(
                    SQLFunction("count", args: SQLLiteral.all),
                    .equal,
                    SQLBind(pairs.count)
                )
                .all()

            imageIDs = rows.compactMap { try? $0.decode(column: "image_id", as: Int.self) }
        }

        // Load the matched images for display.
        var images: [Image] = []
        if !imageIDs.isEmpty {
            images = try await Image.query(on: req.db)
                .filter(\.$id ~~ imageIDs)
                .all()
        }

        // Tag/value counts within the matched set, each combined with the
        // existing filter pairs to build a (deeper) filter hash.
        var refined: [ExifTagCount] = []
        if !imageIDs.isEmpty {
            let counts = try await BrowseHelpers.tagCounts(on: req.db, imageIDs: imageIDs)
            for c in counts where !excludedTags.contains(c.tag) {
                var combined = pairs
                combined.append(TagValue(tag: c.tag, value: c.value))
                let h = FilterHash.hash(combined)
                try await BrowseHelpers.ensureSearch(hash: h, tagValues: combined, on: req.db)
                refined.append(ExifTagCount(tag: c.tag, value: c.value, count: c.count, q: h))
            }
        }

        struct Context: Content {
            let pairs: [TagValue]
            let images: [Image]
            let refined: [ExifTagCount]
        }
        return try await req.view.render(
            "filter/hash_filter",
            Context(pairs: pairs, images: images, refined: refined)
        )
    }
}
